boot/bootutil/src/loader.c

```diff
@@ -1577,6 +1577,201 @@ boot_set_active_image(size_t image_id)
+    rc = flash_area_read(fap, off, &info, sizeof(info));
+    if (rc != 0) {
+        rc = BOOT_EFLASH;
+        goto done;
+    }
+
+    if (info.it_magic != IMAGE_TLV_INFO_MAGIC) {
+        rc = -1;
+        goto done;
+    }
+    end = off + info.it_tlv_tot;
+    off += sizeof(info);
+
+    /* Traverse through all of the TLVs to find the dependency TLVs. */
+    for (; off < end; off += sizeof(tlv) + tlv.it_len) {
+        rc = flash_area_read(fap, off, &tlv, sizeof(tlv));
+        if (rc != 0) {
+             rc = BOOT_EFLASH;
+             goto done;
+         }
+
+        if (tlv.it_type == IMAGE_TLV_DEPENDENCY) {
+            if (!dep_tlvs_found) {
+                dep_tlvs_found = true;
+            }
+
+            if (tlv.it_len != sizeof(dep)) {
+                rc = -1;
+                goto done;
+            }
+
+            rc = flash_area_read(fap, off + sizeof(tlv), &dep, tlv.it_len);
+            if (rc != 0) {
+                rc = BOOT_EFLASH;
+                goto done;
+            }
+
+            /* Verify dependency and modify the swap type if not satisfied. */
+            rc = boot_verify_single_dependency(&dep);
+            if (rc != 0) {
+                /* Dependency not satisfied. */
+                rc = -1;
+                goto done;
+            }
+
+            /* Dependency satisfied, no action needed.
+             * Continue with the next TLV entry.
+             */
+        } else if (dep_tlvs_found) {
+            /* The dependency TLVs are contiguous in the TLV area. If a
+             * dependency had already been found and the last read TLV
+             * has a different type then there are no more dependency TLVs.
+             * The search can be finished.
+             */
+            break;
+        }
+    }
+
+done:
+    flash_area_close(fap);
+    return rc;
+}
+
+/**
+ * Verify whether the image dependencies in the TLV area are
+ * all satisfied and modify the swap type if necessary.
+ *
+ * @return                  0 if all dependencies are satisfied,
+ *                          nonzero otherwise.
+ */
+static int
+boot_verify_single_image_dependency(void)
+{
+    size_t slot;
+    int32_t rc;
+
+    /* Determine the source of the dependency TLVs. Those dependencies have to
+     * be checked which belong to the image that will be located in the primary
+     * slot after the firmware update process.
+     */
+    if (boot_data->swap_type != BOOT_SWAP_TYPE_NONE &&
+        boot_data->swap_type != BOOT_SWAP_TYPE_FAIL) {
+        slot = BOOT_SECONDARY_SLOT;
+    } else {
+        slot = BOOT_PRIMARY_SLOT;
+    }
+
+    rc = boot_verify_all_dependency(slot);
+
+    return rc;
+}
+
+/**
+ * Iterate over all the images and verify whether the image dependencies in the
+ * TLV area are all satisfied and update the related swap type if necessary.
+ */
+static void
+boot_verify_all_image_dependency(void)
+{
+    size_t img_cnt = 0;
+
+    while(img_cnt < BOOT_IMAGE_NUMBER) {
```

Comment:
Also is this routine supposed to hang the bootloader on errors?

Reply:
> Also is this routine supposed to hang the bootloader on errors?

Could you elaborate this please?

Reply:
You can find this part in the current code base at: https://github.com/JuulLabs-OSS/mcuboot/blob/6d079f994aa3ff89a8c8cd85d6f7286487006e91/boot/bootutil/src/loader.c#L1775-L1798 .

Referring to our former conversation:
As far as you could remember (to the previous states of the code), your concern was that it is possible to hang in the loop and never boot an image if any of the upgrades failed. Could you please check the current code regarding to this?

If an image dependency fails, the bootloader refuses the update of that image and in the next cycle it will try to verify the dependencies of the previous version of that image. In worst case the system can return to its previous state (it is assumed that there is always at least one valid state, the initial state - before the bootloader starts - must be consistent).